Implement test environments and before/after closures

regexident opened the pull request:
Hi @mackwic,

consider this the first in a series of PRs aiming to improve rspec, as layed out in tracking issue https://github.com/mackwic/rspec/issues/29.

It **implements a shared context/example environment with mutability isolation**, which aims to **enable sophisticated test scenarios** as well as hopefully allowing us to **safely implement concurrent test execution** without having to worry about environment corruption, thanks to the power of Rust's borrow checker. (as such it aims to fix issue https://github.com/mackwic/rspec/issues/7)

Further more it **adds support for specifying multiple `before_all` (aka `before`), `after_all` (aka `after`) closures** (with guarantees of preserved execution order) to be executed prior/posterior to execution of a context's context or example members. (as such it aims to fix issue https://github.com/mackwic/rspec/issues/20)

Last but not least it **adds support for specifying multiple `before_each`, `after_each` closures** (with guarantees of preserved execution order) to be executed prior/posterior to execution of each of a context's context or example members. (as such it aims to fix issue https://github.com/mackwic/rspec/issues/8)

Please note that during the Runner's complete overhaul I removed its unit tests at some point and haven't yet found the time to add them back. So consider this a WIP, I guess?

Anyway, I'd love to hear your feedback.

Ps: I also upped the Rust version on TravisCI from 1.9 to 1.18, for the sake of being able to use `pub(crate)` and other neat things. I hope that's fine with you?

Pps: Also expect the next PRs to be more finely grained. :wink:

mackwic commented on src/context.rs:
```diff
@@ -1,702 +1,768 @@
 //! Running these tests and doing asserts is not the job of the Context, but the Runner, which is
 //! a struct returned by the root context declaration.
-//!
-//! # Examples
```
replace with `TODO: put updated code example here` ?

regexident replied:
Good point.

mackwic commented on src/example_report.rs:
```diff
@@ -0,0 +1,90 @@
+
+impl Failure {
+    pub fn new<S>(message: Option<S>) -> Self
+        where String: From<S>
```
Where String implements `From<S>` ? Why not `S : ToString` ?

regexident replied:
Sure. To be honest I didn't have `ToString` on my screen at all, when I wrote this. 😇

mackwic replied:
To be frank, it's not the same. There is no blanket impl `String : From<T: ToString>` so it you depends on this bound, it will break something.

Where do you use this ?

regexident replied:
It is used for `Failure` creation from either a `str` (literal) or a `String` (`format!(…)`).

mackwic commented on src/example_report.rs:
```diff
@@ -0,0 +1,90 @@
+}
+
+// #[derive(Clone, PartialEq, Eq, Debug)]
+// pub struct ExampleReport(Result<(), String>);
```
dead code to remove

regexident replied:
Yup.

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,428 @@
-                if result.is_err() && !self.name_stack.is_empty() {
-                    let failure_name = self.name_stack.join(" | ");
-                    self.failures.push(failure_name);
+    fn exit_suite(&mut self, report: &ContextReport) {
```
too much logic here

regexident replied:
I have a second PR almost ready however which completely parallelizes the entire test suite execution (configurable) via *rayon*. It completely messes up the console log for obvious reasons. With this in mind I'd suggest to just ignore this one for now when deciding when and how to merge this PR, as I plan to rework this part (and add unit tests) in the very next PR to follow. But for the sake of clarity I'd rather not any further major refactors to this PR.

The next PR also turns the entire suite/context/example hierarchy immutable and makes the runner execution immutable (as in `…(&self, …)`, plus some `Arc<Mutex<…>>` on the `EventHandler`s) as well. This whole immutability thing is what enables the fearless concurrency _cough_ to begin with.

mackwic replied:
As I said, I am OK with a complete big PR. Sometimes refactoring are large and it's OK.

Why don't you push it all here, and we work the rest together ?
At least please place block comments that signal where the moving code is, but I suspect a change in type might break your next PR...

mackwic commented on src/runner.rs:
```diff
@@ -6,511 +6,183 @@
+    pub fn run(mut self) -> ContextReport {
+        let suite = mem::replace(&mut self.suite, None).expect("Expected context");
+        panic::set_hook(Box::new(|_panic_info| {
+            // silently swallows panics
```
this should fail the test, no ?

regexident replied:
You mean the `// silently swallows panics`? Oh that just prevents the panic from getting printed to the console.

It's [this](https://github.com/mackwic/rspec/pull/30/files/f8be924fbe008473ffd29ec63ea61299043fe360#diff-e0934bb6f70a4f8c3baf22a773c14ec7R414) that does the actual fail catching.

mackwic replied:
Oh ! Okay, so then maybe change the comment to: `// XXX panics already catched at the test call site, we want to discard the trace` ?

mackwic commented on src/runner.rs:
```diff
@@ -6,511 +6,183 @@
-            result = test_res.or(result);
+    pub fn run_or_exit(self) {
+        if self.run().failed > 0 {
+            ::std::process::exit(101);
```
why 101 ?

regexident replied:
I basically checked what `$ cargo test` returns for a failing test (via `$ echo $?`). Turns out it's `101`.

https://www.reddit.com/r/rust/comments/3qdvez/guarantees_about_exit_codes/cwf46cb/

mackwic replied:
Excellent ! I suggest you put this link in comment just before the `process::exit` line. Something like that:
```rust
   pub fn run_or_exit(self) {
        if self.run().failed > 0 {
             // XXX Cargo test failure returns 101. Ref: https://www.reddit.com/r/rust/comments/3qdvez/guarantees_about_exit_codes/cwf46cb/

            ::std::process::exit(101);
```

mackwic commented:
There is _tremendous_ work here, I can see it. I it is impressing. And I quite like the look of it !

this was a quick look at the code, I need to load it in my IDE and play with it a bit to suggest modifications. This weekend I'll make remarks.

regexident commented on .travis.yml:
```diff
@@ -11,7 +11,7 @@ rust:
 - beta
 - stable
-- 1.9.0
+- 1.18.0
```
Cool.

regexident commented on Cargo.toml:
```diff
@@ -1,16 +1,37 @@
+name = "rspec"
+readme = "README.md"
+repository = "https://github.com/mackwic/rspec"
+version = "1.0.0-beta.3"
```
Followed by hopefully quite a bunch of betas and a final RC, yeah.

regexident commented on Cargo.toml:
```diff
@@ -1,16 +1,37 @@
+
+# [dependencies.expectest]
+# optional = true
+# version = "0.8.0"
```
Yeah, this would have to be bumped to `0.9.1` and the other git branch one removed.

regexident commented on Cargo.toml:
```diff
@@ -1,16 +1,37 @@
+[dependencies.expectest]
+optional = true
+git = "https://github.com/regexident/expectest"
+branch = "consistency"
```
Thanks. expectest + rspec are a beautiful match, imho. :wink:

regexident commented on Cargo.toml:
```diff
@@ -1,16 +1,37 @@
 
 [features]
 default = []
+use_expectest = ["expectest"]
```
I like that.

regexident commented on README.md:
```diff
@@ -12,7 +12,7 @@ The last stable documentation is available for consultation at
 
 **All rspec releases are garanteed to compile against the latest stable rust and
-are tested on all rust versions from the 1.9**.
+are tested on all rust versions from the 1.16**.
```
Yes. I first went with 1.16, but then realized I need 1.18 for a thing I can't recall right now.

regexident commented on README.md:
```diff
@@ -42,8 +42,8 @@ fn add(x: u32, y: u32) -> u64 {
 fn test_add() {
-    rdescribe("add", |ctx| {
-        ctx.describe("0 <= x + y <= u32::MAX", |ctx| {
+    rdescribe("add", (), |ctx, _| {
```
Oh, this is a remainder of an intermediate state of the PR. I need to proof-read and update README.

regexident commented on examples/simple.rs:
```diff
@@ -1,40 +1,67 @@
+    let mut runner = rspec::given("a BTreeSet", environment, |ctx| {
+        ctx.when("not having added any items", |ctx| {
+            ctx.then("it is empty", |env| {
+                assert!(env.set.is_empty())
```
Oh, absolutely. I guess I went with `assert!` for the sake of familiarity with rustc's test crate.

regexident commented on src/context.rs:
```diff
@@ -1,702 +1,768 @@
-//! The Context module holds all the functionality for the test declaration, that is: `describe`,
-//! `before`, `after`, `it` and their variants.
+//! The Context module holds all the functionality for the test declaration, that is:
```
Open to it, but not sure what's be the benefit? 🤔

regexident commented:
> There is tremendous work here, I can see it. I it is impressing. And I quite like the look of it !

Awesome, thanks!

> this was a quick look at the code, I need to load it in my IDE and play with it a bit to suggest modifications. This weekend I'll make better remarks.

Cool. As already mentioned in one of the comments I have another PR in the works (as a follow-up for this one) that turns the whole thing into a parallel executed test Runner that walks over a strictly immutable hierarchy of suites/contexts/examples with the only mutable thing being the environment that is being passed around safely and with strict mutability isolation per context.

For all this I need to completely overhaul the formatter. And while I'm doing that I also plan to add back the unit test that I had to remove from runner and formatter.

mackwic commented on src/runner.rs:
```diff
@@ -6,511 +6,183 @@
+            // silently swallows panics
+        }));
+        self.visit(&suite);
+        let _ = panic::take_hook();
```
this is not clear and missed it at first review. Add a comment `// XXX unregister panic hook` ?

regexident commented:
I just pushed the WIP that makes the entire test execution immutable (with exception of the environment that's getting passed around and forked).

It further more adds support for using a single runner with multiple suites, and removes lots of debris from the past, such as unused explicit lifetimes and `mut` markers.

The formatter's output breaks when run in parallel mode with `Configuration::default()` (i.e. `parallel = true`). I hope to find time tomorrow to tackle this by overhauling the hole formatting thing.

Please feel free to point out any other parts that still need some work to get merged.

mackwic commented on src/runner.rs:
```diff
@@ -6,511 +6,180 @@
-        self.handlers.broadcast(&Event::FinishedRunner(result));
-        result
+impl Runner {
+    pub fn run<T>(self, suite: (Suite<T>, T)) -> SuiteReport
```
I don't understand the difference between `run` and `run_with`

mackwic replied:
(same for `run_and_exit` and its `run_and_exit_with` counterpart)

regexident replied:
One takes an explicit configuration, one uses the implicit default.

However given that test initiation has become quite cumbersome with the latest changes it doesn't make much a change, I guess. Maybe just drop the default config one and remove the `_with` from the explicitly configured variant.

My plan is to at some point (i.e. once the overall architecture has settled) re-introduce some API akin to `rdescribe(…)`, which would wrap all this:

```rust
let simple = rspec::formatter::Simple::new(io::stdout());
let formatter = Arc::new(Mutex::new(simple));
let configuration = Configuration::default(); // .parallel(false);
let runner = Runner::new(configuration, vec![formatter]);

#[derive(Clone, Debug)]
struct Environment {
    // …
}

let environment = Environment {
    // …
};

runner.run_or_exit(rspec::given("a BTreeSet", environment, |ctx| {
    // …
}));
```

into something more palatable, like this:

```rust
#[derive(Clone, Debug, Rspec)]
struct Environment {
    // …
}

rspec::run_or_exit(rspec::given("a BTreeSet", Environment {
    // …
}, |ctx| {
    // …
}));
```

mackwic commented:
Are you OK if I push a Rustfmt commit on this branch ?

For the next changes modifications, I'd like to make PRs on this PR.

regexident commented on src/context_member.rs:
```diff
@@ -0,0 +1,11 @@
+}
+
+unsafe impl<T> Send for ContextMember<T> where T: Send {}
+unsafe impl<T> Sync for ContextMember<T> where T: Sync {}
```
Necessary for parallelization. Good point, will document.

regexident commented on src/example.rs:
```diff
@@ -0,0 +1,44 @@
+}
+
+unsafe impl<T> Send for Example<T> where T: Send {}
+unsafe impl<T> Sync for Example<T> where T: Sync {}
```
Necessary for parallelization. Good point, will document.

regexident commented:
> Are you OK if I push a Rustfmt commit on this branch ?

Oh please, do. 🙂

> For the next changes modifications, I'd like to make PRs on this PR.

Fine with me. I don't plan to do any major changes on this one, other than cleanups.
Unless you want me to push the reworked Formatter (to make it work with parallel test runs) here as well, before merging. Up to you. 🙂

I'm currently looking into the concurrent formatter thing.

mackwic commented:
Great !

regexident commented:
So, anything in particular you'd like me to fix prior to merging?

mackwic commented:
Yes, there is a lot of work to be done in this PR before being ready to merge. Documentation, tests, examples are the obvious one. I also saw some bigger changes but I am not sure I understand how it could break the rest (like duplication). I am currently playing with it in my IDE to test changes.

When it's a little thing I can put a comment. But when it's a big change, as I already tested it in my IDE, I make a change proposal via external PRs (eg: #31 and #32 ), all dependants PRs needs to be merged before merging this one. It's like a comment on your diff, but I already tested the code. 

The goal is to keep the debate contained in little changes and improve progressively. You can disagree on my proposal (it's fine !), or have better ideas on how to do it. When we agree, we merge, we move forward, and we have better code. 👌 

Is this fine by you ?

regexident commented:
> Yes, there is a lot of work to be done in this PR before being ready to merge. Documentation, tests, examples are the obvious one. I also saw some bigger changes but I am not sure I understand how it could break the rest (like duplication). I am currently playing with it in my IDE to test changes.

Good point. I will add a couple more examples and extend the existing documentation.

> When it's a little thing I can put a comment. But when it's a big change, as I already tested it in my IDE, I make a change proposal via external PRs (eg: #31 and #32 ), all dependants PRs needs to be merged before merging this one. It's like a comment on your diff, but I already tested the code.

Cool.

> The goal is to keep the debate contained in little changes and improve progressively. You can disagree on my proposal (it's fine !), or have better ideas on how to do it. When we agree, we merge, we move forward, and we have better code. 👌

👍 

> Is this fine by you ?

Absolutely!

regexident commented:
#31 & #32 have been merged successfully.

Rebase and merge this one or are there any remaining things to fix (fine with me, for what it's worth)?

mackwic commented:
:( you merged too fast. I had comment waiting to be posted on #32

mackwic commented:
also I have comments waiting for this one. Why the rush ?

There is almost no test here, and this is a large PR. I am not confident on a merge.

Are you _sure_ you can't make this better ? I see dozen of opportunities here. Starting from doc and examples. Code is no use if untested, undocumented, and not in a state a user can understand easily.

Again, I am sorry I don't have a full time availability to review. I have currently pro and personal obligations that I must oblige first.

mackwic commented:
I'll dump comments here and there so al least you have an idea of what I am thinking

mackwic commented on README.md:
```diff
@@ -42,8 +42,8 @@ fn add(x: u32, y: u32) -> u64 {
-    rdescribe("add", |ctx| {
-        ctx.describe("0 <= x + y <= u32::MAX", |ctx| {
+    rdescribe("add", (), |ctx, _| {
+        ctx.describe("0 <= x + y <= u32::MAX", (), |ctx, _| {
```
- README example needs update
- copy this example in a file in `examples/`
- document this in the readme like "test this with `cargo run --example....`

mackwic commented on README.md:
```diff
@@ -95,17 +97,16 @@ pub fn main() {
     });
-    runner.add_event_handler(&mut formatter);
+    runner.add_event_handler(formatter);
     runner.run().unwrap();
```
update this

mackwic commented on examples/game_of_life.rs:
```diff
@@ -0,0 +1,13 @@
+//!
+//! This Example show how the Game of Life could be implemented and tested with RSpec
+//!
```
this shouldn't have been pushed. Mistake by me. To remove

mackwic commented on examples/given_when_then.rs:
```diff
@@ -1,22 +0,0 @@
-
-extern crate rspec;
```
keep the example

regexident replied:
The [README](https://github.com/rust-rspec/rspec/pull/30/files#diff-04c6e90faac2675aa89e2176d2eec7d8) now contains a paragraph (`### Suites, Contexts & Examples`) on the different styles. Sufficient? No? Comments?

mackwic replied:
Yes the readme is great, but there is never enough small, simple, focused examples. :)

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
-use rspec::context::describe;
+
 use std::io;
+use std::sync::{Arc, Mutex};
```
this is not a simple example. Rename to something else

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
+        ctx.when("adding an new item", |ctx| {
+            ctx.before_all(|env| {
+                env.len_before = env.set.len();
+                env.set.insert(42);
```
comment with explanation

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
+    };
 
-            ctx.it("failed if the return is_err()", || Err(()) as Result<(),()>);
+    runner.run_or_exit(rspec::given("a BTreeSet", environment, |ctx| {
```
comment with explanation

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
+use rspec::prelude::*;
 
-        ctx.it("can define tests", || true);
+pub fn main() {
```
keep in mind examples are meant to be read. This is not readable by a beginner. Please comment as you saw this file for the first time.

What question would ask you a newbie ?

mackwic replied:
also: is this file concise enough ? Straight to the point ? Maybe the purpose f this example is not clear enough and needs to be split to show something precise ?

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
+            });
 
-            ctx.it("should fail if false", || false);
+            ctx.then("its len increases by 1", move |env| {
```
explain move

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
 
-            ctx.it("should pass if true", || true);
+            ctx.then("it is not empty any more", |env| {
+                assert!(!env.set.is_empty());
```
use bool return when it's more readable

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,65 @@
 
-            ctx.it("is convenient for asserts", || assert_eq!(1, 1));
+        ctx.then("panic!(…) fails", move |_env| {
+            panic!("Some reason for failure.")
```
explain the panic usage. It may not be relevant in this file, but in another example

mackwic commented on src/context/mod.rs:
```diff
@@ -0,0 +1,48 @@
+//!
+//! Running these tests and doing asserts is not the job of the Context, but the Runner, which is
+//! a struct returned by the root context declaration.
+//!
```
need more docs + example

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+        T: ::std::fmt::Debug,
+    {
+        let info = name.into().map(|name| {
+            ContextInfo {
```
use a constructor in `ContextInfo`

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+    /// #
+    /// # pub fn main() {
+    /// #     let simple = rspec::formatter::Simple::new(io::stdout());
+    /// #     let formatter = Arc::new(Mutex::new(simple));
```
thought: if the formatter needs `Arc<Mutex<F>`, why not put it in the formatter itself ?

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+        self.context_internal(info, body)
+    }
+
+    /// Open a new name-less context within the current context which won't show up in the logs.
```
what's the use case ?

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+        F: 'static + Fn(&T) -> U,
+        U: Into<ExampleReport>,
+    {
+        let info = ExampleInfo {
```
use constructor

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+                name: name.to_owned(),
+            }
+        });
+        self.context_internal(info, body)
```
thought: a trait would permit a cleaner separation + put the doc in another place than the logic

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+                    let message = error_as_str.or(error_as_string).map(|cow| {
+                        let message = cow.to_string();
+                        format!("thread panicked at '{:?}'.", message)
+                    });
```
extract the string handling elsewhere in a (tested) function

mackwic replied:
this code is not great. How could it be better ?

mackwic commented on src/context/context.rs:
```diff
@@ -0,0 +1,558 @@
+        use std::panic::{catch_unwind, AssertUnwindSafe};
+
+        let test = Example::new(info, move |environment| {
+            let result = catch_unwind(AssertUnwindSafe(|| body(&environment).into()));
```
extract elsewhere in a (tested) function

mackwic commented on src/context/example.rs:
```diff
@@ -0,0 +1,53 @@
+use report::example::ExampleReport;
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum ExampleLabel {
```
not a big fan of this struct.

Thought: use [strum](https://github.com/Peternator7/strum) which does the same via a macro ? Won't make it away, but it's less code

regexident replied:
But is getting rid of 9 clear and readable lines …

```rust
impl From<ExampleLabel> for &'static str {
    fn from(label: ExampleLabel) -> Self {
        match label {
            ExampleLabel::It => "It",
            ExampleLabel::Example => "Example",
            ExampleLabel::Then => "Then",
        }
    }
}
```

… worth the additional dependency and macro magic? With the above you know what's going on. With a macro one would have to it look up.

mackwic replied:
I really don't know. That's why I didn't posted it first. It's a thought.

mackwic commented on src/context/example.rs:
```diff
@@ -0,0 +1,53 @@
+}
+
+pub struct Example<T> {
+    pub(crate) info: ExampleInfo,
```
why sometimes `pub` and sometimes `pub(crate)` ? This needs to be clarified, and made coherent

regexident replied:
`pub(crate)` are used from other modules, but not exposed as public API. I'm quite certain their use is coherent, as in lower the visibility to the minimum of what's possible. I don't want the user to be able to create `Example<T>` at will, that's why `pub(crate) fn new<F>(info: ExampleInfo, f: F) -> Self` is `pub(crate)`, e.g.

mackwic replied:
I mean I believe we can use `pub(crate)` almost everywhere

regexident replied:
Depends on whether we want to change `Event` from `pub` to `pub(crate)`.
Right now one can write one's own custom `EventHandler`.
The `…Info`, and thus all other contained types structs are part of `Event`.

mackwic commented on src/context/mod.rs:
```diff
@@ -0,0 +1,48 @@
+pub use self::example::*;
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum ContextLabel {
```
same as ExampleLabel

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
+use std::io;
+use std::mem;
+
+use colored::*;
```
doesn't work on windows. Needs a TODO

regexident replied:
I'm not experienced with Rust on Windows, I have to admit. Have a suggestion for what to put there?

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
-                report.error_count)
+    fn enter_suite(&mut self, info: &SuiteInfo) {
+        self.name_stack.push(ScopeInfo::Suite(info.clone()));
+        let _ = writeln!(self.buffer, "\nrunning tests");
```
In case of error, as you eat the return, you keep writing in a buffer that doesn't want to be written.

Error stops the function

regexident replied:
Mind to elaborate?

mackwic replied:
If the `writeln!` returns an error, it means that we can't write in the buffer, so we should just stop the function and log an error somewhere.

In the current code, we continue to write even with an error (because it's eaten away), this is not right.

regexident replied:
Ah, now I see. Consider it fixed.

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
-                if result.is_err() && !self.name_stack.is_empty() {
-                    let failure_name = self.name_stack.join(" | ");
-                    self.failures.push(failure_name);
+    fn exit_suite(&mut self, report: &SuiteReport) {
```
too much responsibility in this function.

Refactoring to do:
1) error handling
2) this function call other elements with buffer and current state
3) other elements padd and write in the given buffer in their own function

regexident replied:
Yes. The way `formatter::Simple` (which by now should probably be renamed to `formatter::Standard`, as it's not simple any more) is implemented right now isn't really compatible with parallel execution anyway.

The `name_stack: Vec<ScopeInfo>` should not be managed inside the formatter, but the `Runner` should manage it along with the environments being passed around and then pass the name stack to the formatter by reference via events. I'm on it.

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
+                    ScopeInfo::Context(info) => {
+                        let padding = Self::padding(indent);
+                        let label: &str = info.label.into();
+                        let _ = writeln!(self.buffer, "{}{} {:?}:", padding, label, info.name);
```
I believe you said multiple times the same thing, as there is duplication in this code. I say it because I see twice the same code. :) This needs to be factorized

regexident replied:
There is no formatter `formatter:Simple` anymore. It is called `formatter::serial::Formatter` now and has a corresponding `formatter::parallel::Formatter` equivalent.

The former prints output to the console as events come in, while the latter waits for all tests in a suite to have finished, before printing anything.

We might want to pass the config to the formatter through the `EventHandler` API to allow the implementation of a smart wrapper type that intelligently chooses the proper strategy for a given `Configuration`. We might end up having to do so anyway, as we might want to be able to configure whether to show timing information in the log, or not, etc. 🤔

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
+        let _ = write!(self.buffer, "; {} failed", report.failed);
+        let _ = write!(self.buffer, "; {} ignored", report.ignored);
+        let _ = write!(self.buffer, "; {} measured", report.measured);
+        let _ = writeln!(self.buffer, "");
```
write once when you can

regexident replied:
Sure, but is this …

```rust
let _ = writeln!(self.buffer, "\ntest result: {}. {} passed; {} failed; {} ignored; {} measured", label, report.passed, report.failed, report.ignored, report.measured);
```

… really any better than this …

```rust
let _ = write!(self.buffer, "\ntest result: {}.", label);
let _ = write!(self.buffer, " {} passed", report.passed);
let _ = write!(self.buffer, "; {} failed", report.failed);
let _ = write!(self.buffer, "; {} ignored", report.ignored);
let _ = write!(self.buffer, "; {} measured", report.measured);
let _ = writeln!(self.buffer, "");
```

?

The latter neatly fits in 80 chars, neatly associates each value with its format location, can easily be extended for additional fields, …

And I'm not sure performance should be of higher concern than usability here, if that's what you mean.

mackwic replied:
1) I won't mind report `impl`ing `Display`
2) a line with breaks should do the trick, no ?

```rust
write!(
  self.buffer,
  "\ntest result: {}. {}",
  label, report
).expect("should have written");
```
?

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
+        let label = if report.failed == 0 {
+            "ok".green()
+        } else {
+            "FAILED".red()
```
I think `colored` may be a bit overkill if it's just for this strings

regexident replied:
For just printing `ok`/`FAILED`, absolutely. With more features getting added in future, such as timing and filtering/focussing there will be need for some degree of coloring to ease visual scanning. In other words `ok`/`FAILED` won't be alone for that much longer, I think.

mackwic replied:
ok

mackwic commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
+{
 }
+
+// #[cfg(test)]
```
make tests pass or refactor them

mackwic commented on src/lib.rs:
```diff
@@ -5,22 +5,43 @@
 
-pub mod example_result;
-pub mod events;
+#[cfg(feature = "expectest_compat")]
```
though: need root documentation

mackwic commented on src/report/context.rs:
```diff
@@ -0,0 +1,78 @@
+use std::ops::{Add, AddAssign};
+use std::iter::Sum;
+
+use report::example::ExampleReport;
```
TODO: explain why you implement the `Add` trait

mackwic commented on src/report/context.rs:
```diff
@@ -0,0 +1,78 @@
+use std::ops::{Add, AddAssign};
```
little doc on the use of this module

mackwic commented on src/report/example.rs:
```diff
@@ -0,0 +1,82 @@
+use std::convert::From;
```
little doc here

regexident replied:
Done.

mackwic commented on src/report/example.rs:
```diff
@@ -0,0 +1,82 @@
+use std::convert::From;
```
there is no tests :(

regexident replied:
Now there are.

mackwic commented on src/report/example.rs:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    pub fn is_ok(&self) -> bool {
+        *self == ExampleReport::Success
```
I would prefer `self == &ExampleReport::Success` as it feels more neutral.

With an `*` it's like it's behind a `Box` or a `Mutex`

mackwic commented on src/report/example.rs:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    pub fn is_err(&self) -> bool {
+        *self != ExampleReport::Success
```
same

mackwic commented on src/report/example.rs:
```diff
@@ -0,0 +1,82 @@
+        if other {
+            ExampleReport::Success
+        } else {
+            ExampleReport::Failure(Failure::new(Some("assertion failed: `expected true`")))
```
thought: this message is not helpful

regexident replied:
What else would you put? The `assertion failed: …` is meant to match the format of both rust's native test matchers (`assert!()`), as well as expectest's (as of [this](https://github.com/zummenix/expectest/pull/34) PR).

mackwic replied:
yes I know. It's a thought. Still, we can probably do better like: `expected the condition to be true`

mackwic commented on src/report/mod.rs:
```diff
@@ -0,0 +1,3 @@
+pub mod suite;
```
little doc

mackwic commented on src/report/suite.rs:
```diff
@@ -0,0 +1,3 @@
+use report::context::ContextReport;
```
little doc

mackwic commented on src/runner/configuration.rs:
```diff
@@ -0,0 +1,16 @@
+
```
little doc

regexident replied:
Done.

mackwic commented on src/runner/configuration.rs:
```diff
@@ -0,0 +1,16 @@
+
+impl Configuration {
+    pub fn parallel(self, parallel: bool) -> Self {
+        Configuration { parallel: parallel }
```
`Configuration { parallel : parallel, .. self }`, if possible, is more future-proof

regexident replied:
I'm using the `derive_builder` for Configuration now to keep things clean and future-proof.

After all this method was little more than an attempt at a poor man's builder pattern.

mackwic replied:
yes good idea !

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+//! [`context::rdescribe`](../context/fn.rdescribe.html). You should not try to instanciate
+//! a Runner directly.
+//!
+//! The main methods are `Runner::run` and `Runner::result`.
```
needs update

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+            if let Ok(ref mut handler) = mutex.lock() {
+                handler.handle(&event);
+            } else {
+                println!("Error: lock failed");
```
not ideal :/ Can we do better ?

regexident replied:
No idea either. :/

mackwic replied:
eprintln! + a FIXME for now

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    }
+}
+
+impl<T> Visitor<Suite<T>> for Runner
```
Visitor is a vague name. Can we find a better name ?

regexident replied:
`TestSuiteVisitor<T>` maybe?

mackwic replied:
yes

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+
+impl<T> Visitor<Context<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
```
why debug ?

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    fn visit(&self, context: &Context<T>, environment: &mut Self::Environment) -> Self::Output {
+        if let Some(ref info) = context.info {
+            self.broadcast(Event::EnterContext(info.clone()));
+        }
```
could these 3 lines be extracted elsewhere ?

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, context: &Context<T>, environment: &mut Self::Environment) -> Self::Output {
```
too much logic here

regexident replied:
Consider this fixed.

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+                    after_each_function(&mut environment);
+                }
+                report
+        };
```
extract the closure

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+                    process_member(self, member, environment.clone())
+                })
+                .sum()
+        };
```
It needs to extract the if 

Can the iterator creation be extracted in a function like that:

```rust
let members = self.get_members(); // an iterator that may be parallel depending on the conf
let report = self.process_all_members(members, environment); // the map(process_member) + sum
```

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+        }
+        if let Some(_) = context.info {
+            self.broadcast(Event::ExitContext(report.clone()));
+        }
```
can we extract that elsewhere ?

mackwic commented on src/suite.rs:
```diff
@@ -0,0 +1,145 @@
+use context::Context;
```
little doc

mackwic commented on src/suite.rs:
```diff
@@ -0,0 +1,145 @@
+    F: FnOnce(&mut Context<T>) -> (),
+    T: Clone + ::std::fmt::Debug,
+{
+    // Note: root context's info get's ignored.
```
why ?

mackwic replied:
why ?

regexident replied:
This has long been improved. Comment has been removed.

mackwic commented on src/visitor.rs:
```diff
@@ -0,0 +1,6 @@
+pub trait Visitor<T> {
```
needs doc

mackwic commented:
I hope the rain of comment won't discourage you. I was retaining them to help you with code and suggestions.

Please keep your enthusiasm. This is a great PR with great work. It just needs the extra mile

regexident commented on src/formatter/simple.rs:
```diff
@@ -1,361 +1,433 @@
+                    ScopeInfo::Context(info) => {
+                        let padding = Self::padding(indent);
+                        let label: &str = info.label.into();
+                        let _ = writeln!(self.buffer, "{}{} {:?}:", padding, label, info.name);
```
Well, both of these cases deal with different types. One could use their similarity to abstract them with a trait, but I'm not sure one would gain much clarity from it, let along brevity. Sometimes duplication is the clearer solution.

mackwic replied:
```rust
ScopeInfo::Context(info) => write_context_info(buffer, indent, info)
```
at least

regexident replied:
Deduplicating the whole thing right now.

regexident commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,190 @@
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, context: &Context<T>, environment: &mut Self::Environment) -> Self::Output {
```
Yeah. I've been trying to split things up, but most attempts turned out to be quite involved and often less readable. I'll keep digging though.

regexident commented:
> :( you merged too fast. I had comment waiting to be posted on #32

I'm so sorry. I missed the "It's like a comment on your diff, but I already tested the code." and thought you were just asking for my review/go.

> also I have comments waiting for this one. Why the rush ?

There is no rush. :)

> There is almost no test here, and this is a large PR. I am not confident on a merge.

I'm adding tests to all places I see appropriate right now.

> Are you sure you can't make this better ?

I'm sure I can. I'm on it.

> I see dozen of opportunities here. Starting from doc and examples. Code is no use if untested, undocumented, and not in a state a user can understand easily.

Fair point.

> Again, I am sorry I don't have a full time availability to review. I have currently pro and personal obligations that I must oblige first.

No problem.

> I'll dump comments here and there so al least you have an idea of what I am thinking

Yes, please. Much appreciated. :)

> I am not happy with my comments. They are short, some are not actionable, and they may be viewed as rude.
> 
> Sorry for that, at least they are shared with you.

No worries.

> I hope the rain of comment won't discourage you. I was retaining them to help you with code and suggestions.

Not at all. The more, the merrier.

> Please keep your enthusiasm. This is a great PR with great work. It just needs the extra mile

No worries. I'm preparing things for proper parallel handling (printing, etc.) and overall simplifying things right now.

regexident commented:
Just pushed another big rework.

Doesn't yet address all of your comments and lots of tests are still missing, too.
But I'm on it and getting there, slowly.

With this latest update the code is now in a state where I feel confident enough to add thorough tests without having to rewrite major parts of them over and over for each further refactor.

I also added a convenience macro (`/examples/macro.rs`) and made the formatter compatible with parallel execution. There is a serial and a parallel formatter, strictly speaking. The latter makes use of the former, internally, in order to avoid duplication. It waits for the `exit_suite` event and then passes a replay of all previous events to its internal serial formatter.
Due to this "buffering" [sic!] it may take seconds, if not minutes (depending on the size of your test suite) for actual test results to be printed to stdout.
It thus might be a good idea to add a progress bar as proposed in https://github.com/rust-rspec/rspec/issues/15#issuecomment-323387086. 🤔

regexident commented:
Just pushed a bunch of minimal-effort fixes addressing your review.

There still is much need for tests though. I'll tackle these next.

In the meanwhile I'd love to get your feedback (no feelings hurt) on the status quo.

mackwic commented on src/runner/observer.rs:
```diff
@@ -1,19 +1,18 @@
-    fn enter_example(&self, example: &ExampleHeader) -> io::Result<()>;
-    fn exit_example(&self, example: &ExampleHeader, report: &ExampleReport) -> io::Result<()>;
+/// `RunnerObserver`s can be attached to a [`Runner`](../runner/struct.Runner.html) to observe a
+pub trait RunnerObserver: Send + Sync {
```
A default empty impl for each would be nice, here

mackwic replied:
it would remove a lot of boilerplate code

regexident replied:
Yeah, good idea!

mackwic commented on src/formatter/parallel.rs:
```diff
@@ -26,75 +26,66 @@ where
-        self.serial.exit_suite(suite, report)?;
-
-        Ok(())
+    fn replay_suite(&self, suite: &SuiteHeader, report: &SuiteReport) {
```
I need to think about this...

regexident replied:
It's the cleanest solution I could come up with, that is guaranteed to generate same output for both serial and parallel formatters, yet manages to share the logic.

For the parallel formatter I'm thinking about displaying a progress bar via the **indicatif** crate, while the tests run. Replacing them with the actual (replayed) logs once the suite exits.

mackwic replied:
yes, it's a good idea. I was thinking about making the replay_* methods part of a public trait with empty default impl. Even sequential reporters could beneficiate from it (like a summary of the failures at the end).

Also, big thumb up 👍 . I really like the look of it !

mackwic commented on src/formatter/parallel.rs:
```diff
@@ -3,12 +3,12 @@ use std::io;
 use header::ContextHeader;
 use header::ExampleHeader;
-use event_handler::EventHandler;
+use runner::RunnerObserver;
```
good renaming 👍

mackwic commented:
I'll do a big review today (and maybe tomorrow) and help as much a I can

regexident commented:
> I'll do a big review today (and maybe tomorrow) and help as much a I can

Great! I'm working on getting the test coverage up to where it was, right now. :)

mackwic commented on src/runner/observer.rs:
```diff
@@ -16,3 +16,18 @@ pub trait RunnerObserver: Send + Sync {
     fn exit_example(&self, example: &ExampleHeader, report: &ExampleReport);
 }
+
+#[cfg(feature = "specialization")]
```
that was not what I meant. You can make a default impl for a trait like that:
```rust
trait Hello {
  fn default_int(&self) -> u32 { 24 }
}
struct A;
struct B;

impl Hello for A {}
impl Hello for B {
  fn default_int(&self) -> u32 { 42 }
}
```
It's standard and stable :)

regexident replied:
Fixed.

mackwic commented on src/header/context.rs:
```diff
@@ -12,14 +10,14 @@ pub enum ContextLabel {
-            ContextLabel::Specify => "Specify",
-            ContextLabel::Given => "Given",
-            ContextLabel::When => "When",
+impl fmt::Display for ContextLabel {
```
This is not blocking, just a thought about design.

When I do it, I like to:
1) keep the `From<*Label> for &str`
2) unit test the From<Label> to &str
3) Display is a simple call to `write!(f, "{}", str::from(label))` as IO is nasty for testing (no real problem with this particular case. Just a general remark)

Again, not bloacking.

regexident replied:
Yeah. I couldn't think of a single use case where one might want to access the label as string representation, other than for the internal use for logging, which already makes use of `Display`/`Debug`. Also with the method being dead-simple other than pure test-coverage for the sake of test coverage I couldn't find a good reason to make things more complicated than necessary.

mackwic replied:
good for me ! 👍

mackwic commented:
(comments triage)

mackwic commented:
(the old review is outdated)

mackwic commented on README.md:
```diff
@@ -32,90 +32,131 @@ extern crate rspec;
+    #[derive(Clone, Default, Debug)]
+    struct Environment {
+        // ...
+    }
```
I think the Environment is an advanced feature that should not be used by default for every test. Maybe split the example in 2, one really simple, and another that showcase the Environment use ?

mackwic commented on README.md:
```diff
@@ -32,90 +32,131 @@ extern crate rspec;
+
 The last stable documentation is available for consultation at
-[mackwic.github.io/rspec](https://mackwic.github.io/rspec).
+[https://docs.rs/rspec](https://docs.rs/rspec).
```
the new Readme is quite great ! 👌

mackwic commented on examples/macro.rs:
```diff
@@ -0,0 +1,57 @@
+#[macro_use(rspec_run)]
```
please comment the objective of this example, what it does, why it's great

mackwic replied:
maybe rename the example ? Macro is not clear what it does. With a good name, it may not need a top comment.

I think it a great showcase of the environment...

mackwic commented on examples/multi.rs:
```diff
@@ -0,0 +1,41 @@
+use std::sync::Arc;
+use std::collections::BTreeSet;
+
+// An example of a single runner running multiple test suites in succession.
```
👍

mackwic commented on examples/multi.rs:
```diff
@@ -0,0 +1,41 @@
+    #[derive(Clone, Debug)]
+    struct Environment {
+        set: BTreeSet<usize>,
+    }
```
don't inject an Environment if not needed, keep examples focused ;)

regexident commented:
I'm having a hard time implementing tests for mocking the runner.

This is the farthest I've come so far (add it to `/runner/mod.rs`):
https://gist.github.com/regexident/02ee175fbbac99addea377e38e8f0343
If you wanna give it a try yourself, I'm gonna call it a day and run some errands.

mackwic commented on examples/simple.rs:
```diff
@@ -1,40 +1,58 @@
-
 extern crate rspec;
-use rspec::context::describe;
+
```
this is basically the same as `macro`. We can remove this one

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+//! The Context module holds all the functionality for the test declaration, that is:
```
functionalities

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+//! The Context module holds all the functionality for the test declaration, that is:
+//! `before`, `after`, `suite`, `context`, `it` and their variants.
+//!
+//! A Context can also holds reference to children Contextes, for whom the before closures will be
```
- from whom
- `_before_` (it's a special word)
- s/closures/functions would be clearer for the newcomer

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+//! respect the order of declaration of theses tests.
+//!
+//! Running these tests and doing asserts is not the job of the Context, but the Runner.
+//!
```
👍

regexident replied:
This should mention that for parallel execution the order is undefined, for obvious reasons. Consider it fixed.

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+use header::{ContextLabel, ContextHeader, ExampleLabel, ExampleHeader};
+use report::ExampleReport;
+
+/// Test contexts are a convenient tool for adding structure and code sharing to a test suite.
```
Test Context holds examples, before and after functions, and have a name

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+    pub fn num_examples(&self) -> usize {
+        self.blocks.iter().fold(0, |count, block| {
+            count + block.num_examples()
+        })
```
```rust
self.blocks
      .map(|b| b.num_examples())
      .sum()
```

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+    {
+        // let existing_infos = self.blocks.iter().map(|m| m.header );
+        // let () = existing_infos;
+        // if self.blocks.contain
```
remove dead code

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+            let result = catch_unwind(AssertUnwindSafe(|| body(&environment).into()));
+            match result {
+                Ok(result) => result,
+                Err(error) => {
```
extract the casting of error to a string `message` in another function

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    /// Declares a closure that will be executed once before any
+    /// of the context's context or example blocks are being executed.
```
context's children

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+{
+    /// Open and name a new context within the current context.
+    ///
+    /// Note that the order of execution is not guaranteed to match the declaration order.
```
`**is not**`

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+
+    /// Open and name a new example within the current context.
+    ///
+    /// Note that the order of execution is not guaranteed to match the declaration order.
```
`**is not**`

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    /// Declares a closure that will be executed once before each
+    /// of the context's context or example blocks.
```
context's children

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    /// Declares a closure that will be executed once after any
+    /// of the context's context or example blocks have been executed.
```
context's children

mackwic commented on src/block/context.rs:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    /// Declares a closure that will be executed once after each
+    /// of the context's context or example blocks.
```
context's children

mackwic commented on src/block/suite.rs:
```diff
@@ -0,0 +1,178 @@
+    }
+}
+
+unsafe impl<T> Send for Suite<T>
```
move the Send and Sync impl after the documented block code

regexident replied:
Sure? `Send` and `Sync` belong to `Suite<T>`. The documented code block contains free functions.

mackwic commented on src/formatter/parallel.rs:
```diff
@@ -0,0 +1,74 @@
+    }
+}
+
+impl<T: io::Write> RunnerObserver for ParallelFormatter<T>
```
remove the empty methods now that we have a default empty impl for the `RunnerObserver`

mackwic commented on src/formatter/serial.rs:
```diff
@@ -0,0 +1,501 @@
+        if let Ok(ref mut mutex_guard) = self.state.lock() {
+            let result = f(mutex_guard.deref_mut());
+            if let Err(error) = result {
+                eprintln!("\n{}: {:?}", "error".red().bold(), error);
```
add a `TODO: better error handling`

mackwic commented on src/formatter/serial.rs:
```diff
@@ -0,0 +1,501 @@
+                eprintln!("\n{}: {:?}", "error".red().bold(), error);
+            }
+        } else {
+            eprintln!(
```
add a TODO: better error handling

mackwic commented on src/formatter/serial.rs:
```diff
@@ -0,0 +1,501 @@
+    }
+}
+
+// #[cfg(test)]
```
remove dead code

mackwic commented on src/header/context.rs:
```diff
@@ -0,0 +1,76 @@
+use std::fmt;
+
+/// A [`Context`](../block/struct.Context.html)'s cosmetic label.
```
```
How the [`Context`](../block/struct.Context.html) will be printed by the `Formatter`
```

mackwic commented on src/header/context.rs:
```diff
@@ -0,0 +1,76 @@
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct ContextHeader {
+    pub label: ContextLabel,
+    pub name: &'static str,
```
question: shouldn't this be a `String` ? With this type bound you constrain how the user can declare suites

regexident replied:
It certainly would be more flexible. But then again I'm not sure I'd consider dynamically generated tests (i.e. at runtime) a good practice to begin with. Tests should be K.I.S.S., no? I wouldn't want to have to search for the piece of code that dynamically generated the test case that causes my suite to fail.

regexident replied:
This is not to say I'd not be fine with changing labels back to being `String`.

mackwic commented on src/header/context.rs:
```diff
@@ -0,0 +1,76 @@
+}
+
+impl ContextHeader {
+    pub fn new(label: ContextLabel, name: &'static str) -> Self {
```
derive_new opportunity

mackwic commented on src/block/suite.rs:
```diff
@@ -0,0 +1,178 @@
+}
+
+impl<T> Suite<T> {
+    pub(crate) fn new(header: SuiteHeader, environment: T, context: Context<T>) -> Self {
```
derive_new opportunity

mackwic commented on src/formatter/serial.rs:
```diff
@@ -0,0 +1,501 @@
+}
+
+impl<T: io::Write> SerialFormatterState<T> {
+    pub fn new(buffer: T) -> Self {
```
derive_new opportunity

mackwic commented on src/header/example.rs:
```diff
@@ -0,0 +1,69 @@
+}
+
+impl ExampleHeader {
+    pub fn new(label: ExampleLabel, name: &'static str) -> Self {
```
derive_new opportunity

mackwic commented on src/header/suite.rs:
```diff
@@ -0,0 +1,69 @@
+}
+
+impl SuiteHeader {
+    pub fn new(label: SuiteLabel, name: &'static str) -> Self {
```
derive_new opportunity

mackwic commented on src/report/context.rs:
```diff
@@ -0,0 +1,54 @@
+}
+
+impl ContextReport {
+    pub fn new(sub_reports: Vec<BlockReport>) -> Self {
```
derive_new opportunity

mackwic commented on src/report/suite.rs:
```diff
@@ -0,0 +1,53 @@
+}
+
+impl SuiteReport {
+    pub fn new(header: SuiteHeader, context: ContextReport) -> Self {
```
derive_new opportunity

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,251 @@
+}
+
+impl Runner {
+    pub fn new(configuration: Configuration, observers: Vec<Arc<RunnerObserver>>) -> Runner {
```
derive_new opportunity

mackwic commented on src/lib.rs:
```diff
@@ -5,22 +5,80 @@
-pub mod events;
-pub mod context;
+#[macro_use]
+extern crate derive_builder;
```
we may want to use [derive_new](https://github.com/nrc/derive-new) for all the `fn new()` boilerplate here and there

mackwic replied:
I have marked the opportunities

mackwic commented on src/report/context.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+}
+
+impl Report for ContextReport {
```
tests would be nice here

mackwic replied:
I prototyped a bit, it needs a lot of setup code...
I think it would be nice to make an evolution of `BlockReport`, `ContextReport`, and `ExampleReport` to have fixtures generators.

I would suggest something like:
```rust
// arrange
let report1 = BlockReport::fixture_from(ExampleReport::fixture_failed());
let report2 = BlockReport::fixture_from(ExampleReport::fixture_success());
let context_report = ContextReport::new(vec!(report1, report2));
// assert
assert_eq!(false, context_report.is_success())
```
With all the `fixture_*` functions members that are gated by a `#[cfg(test)]`

That's only a suggestion, ofc

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,251 @@
+        let report = self.visit(&suite, &mut environment);
+        self.clean_after_run();
+        if let Ok(mut mutex_guard) = self.should_exit.lock() {
+            *mutex_guard.deref_mut().get_mut() |= report.is_failure();
```
too many things done in 1 line

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,251 @@
+        Runner {
+            configuration: configuration,
+            observers: observers,
+            should_exit: Mutex::new(Cell::new(false)),
```
why do you need a Mutex<Cell> ?

regexident replied:
I'm not sure we do. I tried lots of things and `Mutex` was the first to make rustc shut up. 😅🙈

Please feel free to improve this! I haven't done too much Rust with mutable state shared across threads. As such I probably chose the wrong types. It also feels way more complicated than it should be.

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,251 @@
+        F: FnMut(&RunnerObserver),
+    {
+        for observer in &self.observers {
+            f(observer.borrow());
```
what's `f` ? Change the name to something explicit please

mackwic replied:
event ?

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,251 @@
+        for before_function in context.before_all.iter() {
+            before_function(environment);
+        }
+        let result = f(environment);
```
same, `f` is not a good name

mackwic replied:
run_function ?

mackwic commented on src/runner/mod.rs:
```diff
@@ -0,0 +1,251 @@
+        for before_function in context.before_each.iter() {
+            before_function(environment);
+        }
+        let result = f(environment);
```
f is not an explicit name

mackwic replied:
visitor_function ?

mackwic commented:
Okay, I'll see tomorrow what I can do

mackwic commented on src/runner/mod.rs:
```diff
@@ -6,12 +6,12 @@ mod observer;
 use std::ops::{Deref, DerefMut};
+use std::panic;
+use std::process;
+use std::sync::{Arc, Mutex};
```
sort the use by alphabetic order for readability and maintenance ease

mackwic commented:
Some theoretical remarks:
- Runner is a pure dispatcher (and this is great !), it doesn't have state per se
- So we only need to test the order of the dispatch and the eventual logic behind (like conditional dispatch)
- You have a hard time to test it because you test too many things at once. When it will break, what would be the message ? It's easier when you do smaller, focused tests
- It's cool the API is generic, I like doing impl for testing needs and inject my little stubs :)

I am impressed by the overall design. It's good, really.

mackwic commented on src/runner/mod.rs:
```diff
@@ -594,4 +594,21 @@ mod tests {
+        fn it_can_be_called() {
+            // arrange
+            let runner = Runner::default();
+            let block = Block::Example(Example::new(ExampleHeader::default(), |_| ExampleReport::Success));
```
seems like a bug to me (or at least a regression).
This wont compile:
```rust
let block = Block::Example(Example::new(ExampleHeader::default(), |_| ()));
```
It should !

mackwic replied:
Same for `false`:
```
/Users/thomas/.cargo/bin/cargo test --color=always --package rspec --lib runner::tests::impl_visitor_block_for_runner -- --nocapture
   Compiling rspec v1.0.0-beta.4 (file:///Users/thomas/projects/rust/rspec)
error[E0308]: mismatched types
   --> src/runner/mod.rs:607:83
    |
607 |             let block = Block::Example(Example::new(ExampleHeader::default(), |_| false));
    |                                                                                   ^^^^^ expected enum `report::example::ExampleReport`, found bool
    |
    = note: expected type `report::example::ExampleReport`
               found type `bool`
```

regexident replied:
No, this is correct. `Example::new(…)` takes a closure which handles the conversion from `()`/`bool`/… to `ExampleReport`. One could use `F: 'static + Fn(&T) -> impl Into<ExampleReport>` instead of `F: 'static + Fn(&T) -> ExampleReport` in future, but that's not available yet. Also `Example::new(…)` is not intended for public use anyway. That's what `ctx.then(…)` is for.

mackwic replied:
okay I see

regexident replied:
However if you see a way to make this cleaner I'm all for it. :)

regexident commented:
> Some theoretical remarks:

> - Runner is a pure dispatcher (and this is great !), it doesn't have state per se

Yeah, this was one of the main goals of my overall concept. `Suite`/`Context`/`Example` are pure, too.

> - So we only need to test the order of the dispatch and the eventual logic behind (like conditional dispatch)

Yes.

> - I believe your test do too many things at once. I find that simpler, little focused tests seems easier to write, maintain, and read IMO

True. Your tests look good! 👍 

> - It's cool the API is generic, I like doing impl for testing needs and inject my little stubs :)

😎

> I am impressed by the overall design. It's good, really.

I'm glad you like it!

I already have timing implemented (yet will have to rebase it and probably rewrite some parts of it now). And filtering/focussing shouldn't be too difficult to implement now either.

regexident commented:
With 97cab2b `Logger` has now reached a state where I'm happy with both implementation and API use.

That being said I plan to add realtime progress for parallel execution via the [indicatif](https://github.com/mitsuhiko/indicatif) crate with a later, separate PR.

---

What is your take on the progress of this PR? What's still missing? What should I fix?

mackwic commented:
LGTM ! I think this PR is now in an acceptable state :)

regexident commented:
Great! Then feel free to rebase and merge it at your discretion.

mackwic commented:
🎉

regexident commented:
Cool, but sure you want this massive PR to be squashed into a single commit?

I think for a massive change of API/implementation like this it would be more than useful to have a more finely-grained log of changes, no? It would also give a better picture of the activity of this project, which with a single commit might give the false impression of still being in some kind of stagnation.

mackwic commented:
I don't want to optimize by the number of commits. It's true, some people look at it but it's a meaningless metric.

I want this PR to be squashed so that we can't checkout an intermediate state from master.

regexident commented:
> I don't want to optimize by the number of commits. It's true, some people look at it but it's a meaningless metric.

Fair point. I personally don't care either.

> I want this PR to be squashed so that we can't checkout an intermediate state from master.

Ah I see, this makes sense.
